### PR TITLE
Hello world plugin added

### PR DESCRIPTION
**Pull request by addy1997**
I have added the hello_world plugin

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

### CODE REVIEW

**Review by chapulina — CHANGES_REQUESTED** [review]
Thank you for the PR. I don't think this is being compiled right now, because there's no `CMakeLists` file.

This also seems to be very similar to [SampleSystem](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/plugin/system_plugin/SampleSystem.cc). Maybe that example can be improved instead of adding another one? What are you envisioning `Hello_world` to help with that isn't accomplished by `SampleSystem`?

**Comment by addy1997** [comment]
> Thank you for the PR. I don't think this is being compiled right now, because there's no `CMakeLists` file.
> 
> This also seems to be very similar to [SampleSystem](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/plugin/system_plugin/SampleSystem.cc). Maybe that example can be improved instead of adding another one? What are you envisioning `Hello_world` to help with that isn't accomplished by `SampleSystem`?

I will add the CMakeLists.txt file right away. Does my code follow the ignition gazebo syntax?

**Comment by addy1997** [comment]
> Thank you for the PR. I don't think this is being compiled right now, because there's no `CMakeLists` file.
> 
> This also seems to be very similar to [SampleSystem](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/plugin/system_plugin/SampleSystem.cc). Maybe that example can be improved instead of adding another one? What are you envisioning `Hello_world` to help with that isn't accomplished by `SampleSystem`?

Well, my only intention is to add a proper Hello world file just like the one in Gazebo classic.

File: examples/plugin/Hello_world/Hello_world.cc
Comment (chapulina) [inline examples/plugin/Hello_world/Hello_world.cc:1]: Please add the copyright header

File: examples/plugin/Hello_world/Hello_world.cc
Comment (chapulina) [inline examples/plugin/Hello_world/Hello_world.cc:11]: Can you update this to explain the current implementation?

File: examples/plugin/Hello_world/Hello_world.cc
Comment (chapulina) [inline examples/plugin/Hello_world/Hello_world.cc:13]: According to our [style guide](https://ignitionrobotics.org/docs/all/contributing#style-guides), all classes must be `CamelCases`, i.e.:

```suggestion
class ignition::gazebo::systems::HelloWorld
```

File: examples/plugin/Hello_world/Hello_world.cc
Comment (chapulina) [inline examples/plugin/Hello_world/Hello_world.cc:32]: Use 2-space indentation

```suggestion
  printf("Hello World!\n");
```

**Review by chapulina — CHANGES_REQUESTED** [review]
> Well, my only intention is to add a proper Hello world file just like the one in Gazebo classic.

Got it, you're talking about this one, right?

https://github.com/osrf/gazebo/tree/gazebo11/examples/plugins/hello_world

> I have added the CMakeLists.txt file.

Our CI is failing because this plugin can't be compiled. Have you tried compiling it locally?

Can you also add a `README` to the plugin's folder like the newest example plugins have? See [this](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/examples/plugin/gui_system_plugin) for example. While working on the README, you should also follow the steps to check that the plugin works for you.

<details>
  <summary>Current build failures</summary>
  
```
2021-03-22T20:24:36.7352330Z [ 50%] Building CXX object CMakeFiles/Hello_world.dir/Hello_world.cc.o
2021-03-22T20:24:36.7353488Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:6:19: error: expected '{' before ';' token
2021-03-22T20:24:36.7354229Z  namespace ignition;
2021-03-22T20:24:36.7354624Z                    ^
2021-03-22T20:24:36.7355519Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:7:17: error: expected '{' before ';' token
2021-03-22T20:24:36.7356257Z  namespace gazebo;
2021-03-22T20:24:36.7356625Z                  ^
2021-03-22T20:24:36.7357527Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:8:18: error: expected '{' before ';' token
2021-03-22T20:24:36.7358253Z  namespace systems;
2021-03-22T20:24:36.7358638Z                   ^
2021-03-22T20:24:36.7359622Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:15:1: error: qualified name does not name a class before '{' token
2021-03-22T20:24:36.7360344Z  {
2021-03-22T20:24:36.7360640Z  ^
2021-03-22T20:24:36.7361526Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:18:1: error: 'Hello_world' does not name a type
2021-03-22T20:24:36.7362570Z  Hello_world::Hello_world() : System(), dataPtr(std::make_unique<Hello_world>())
2021-03-22T20:24:36.7363106Z  ^~~~~~~~~~~
2021-03-22T20:24:36.7364162Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:26:6: error: 'MyPlugin' has not been declared
2021-03-22T20:24:36.7365051Z  void MyPlugin::Configure(const Entity &_entity,
2021-03-22T20:24:36.7365529Z       ^~~~~~~~
2021-03-22T20:24:36.7366443Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:26:32: error: 'Entity' does not name a type
2021-03-22T20:24:36.7367276Z  void MyPlugin::Configure(const Entity &_entity,
2021-03-22T20:24:36.7367764Z                                 ^~~~~~
2021-03-22T20:24:36.7368842Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:28:5: error: 'EntityComponentManager' has not been declared
2021-03-22T20:24:36.7369861Z      EntityComponentManager &_ecm,
2021-03-22T20:24:36.7370360Z      ^~~~~~~~~~~~~~~~~~~~~~
2021-03-22T20:24:36.7371326Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:29:5: error: 'EventManager' has not been declared
2021-03-22T20:24:36.7372152Z      EventManager &/*_eventMgr*/)
2021-03-22T20:24:36.7372580Z      ^~~~~~~~~~~~
2021-03-22T20:24:36.7373707Z /github/workspace/examples/plugin/Hello_world/Hello_world.cc:37:20: error: expected constructor, destructor, or type conversion before '(' token
2021-03-22T20:24:36.7374890Z  IGNITION_ADD_PLUGIN(Hello_world, System, Hello_world::ISystemConfigure)
2021-03-22T20:24:36.7375489Z                     ^
2021-03-22T20:24:36.7376482Z CMakeFiles/Hello_world.dir/build.make:62: recipe for target 'CMakeFiles/Hello_world.dir/Hello_world.cc.o' failed
2021-03-22T20:24:36.7377700Z make[3]: *** [CMakeFiles/Hello_world.dir/Hello_world.cc.o] Error 1
2021-03-22T20:24:36.7378796Z CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/Hello_world.dir/all' failed
2021-03-22T20:24:36.7379596Z make[2]: *** [CMakeFiles/Hello_world.dir/all] Error 2
2021-03-22T20:24:36.7381310Z Makefile:83: recipe for target 'all' failed
2021-03-22T20:24:36.7382003Z make[1]: *** [all] Error 2
2021-03-22T20:24:36.7383015Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/command_actor[0m[1;36m[0m
2021-03-22T20:24:36.7384175Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/oTA1RF[0m[1;36m[0m
2021-03-22T20:24:36.7385351Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/custom_component[0m[1;36m[0m
2021-03-22T20:24:36.7386513Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/YQWpPd[0m[1;36m[0m
2021-03-22T20:24:36.7387677Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/gui_system_plugin[0m[1;36m[0m
2021-03-22T20:24:36.7388822Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/30O5P9[0m[1;36m[0m
2021-03-22T20:24:36.7389968Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/system_plugin[0m[1;36m[0m
2021-03-22T20:24:36.7391123Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/LUwUwT[0m[1;36m[0m
2021-03-22T20:24:36.7392267Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/Hello_world[0m[1;36m[0m
2021-03-22T20:24:36.7395745Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/WykYGa[0m[1;36m[0m
2021-03-22T20:24:36.7396550Z /github/workspace/test/integration/examples_build.cc:164: Failure
2021-03-22T20:24:36.7397300Z Expected equality of these values:
2021-03-22T20:24:36.7397759Z   system(cmd)
2021-03-22T20:24:36.7398107Z     Which is: 512
2021-03-22T20:24:36.7398431Z   0
2021-03-22T20:24:36.7398745Z Hello_world
2021-03-22T20:24:36.7399392Z [  FAILED  ] Plugins/ExamplesBuild.Build/0, where GetParam() = "plugin" (56687 ms)
```

</details>

**Comment by addy1997** [comment]
> > Well, my only intention is to add a proper Hello world file just like the one in Gazebo classic.
> 
> Got it, you're talking about this one, right?
> 
> https://github.com/osrf/gazebo/tree/gazebo11/examples/plugins/hello_world
> 
> > I have added the CMakeLists.txt file.
> 
> Our CI is failing because this plugin can't be compiled. Have you tried compiling it locally?
> 
> Can you also add a `README` to the plugin's folder like the newest example plugins have? See [this](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/examples/plugin/gui_system_plugin) for example. While working on the README, you should also follow the steps to check that the plugin works for you.
> Current build failures


Yes, this plugin. No, I couldn't manage to compile it locally cause I got this error` Unknown CMake command "ign_find_package"`. Also, I asked this on the [Discord](https://answers.gazebosim.org/question/26770/unknown-cmake-command-ign_find_package/).

**Comment by addy1997** [comment]
@chapulina Ma'am, it was not slightest of my intention to waste your time(). I have updated the code and have a compiled plugin (terminal output below) for your reference. 

##Terminal Output
```

adwaitnaik@ubuntu:~/plugin_ws/src/hello_world/build$ make 
Scanning dependencies of target HelloWorld
[ 50%] Building CXX object CMakeFiles/HelloWorld.dir/HelloWorld.cc.o
[100%] Linking CXX shared library libHelloWorld.so
[100%] Built target HelloWorld

```
I could compile it locally.  Lastly, here's the [library](https://github.com/addy1997/ign-gazebo/tree/Hello_world_plugin/examples/plugin/Hello_world/lib). 

Thanks

**Comment by chapulina** [comment]
> it was not the slightest of my intention to waste your time

Sorry if that's what it sounded like! I completely understand you have the best intentions, I was just making suggestions on how to optimize the review process :wink: This is in the review queue :+1:

**Comment by addy1997** [comment]
@chapulina, any update on this?

**Comment by chapulina** [comment]
Hi @addy1997 , the plugin still can't be compiled. I see you're using ROS packages like catkin, roscpp and rospy, which aren't dependencies of Ignition. It may compile on your system if you have those on your workspace, but to be included into our examples, it can't depend on ROS.

You can see how the test fails to compile on our CI. Click on the "Details" for the "Ubuntu CI" check and see on the logs how your plugin fails to compile:

```
2021-04-06T16:56:24.0205669Z CMake Error at CMakeLists.txt:10 (find_package):
2021-04-06T16:56:24.0206438Z   By not providing "Findcatkin.cmake" in CMAKE_MODULE_PATH this project has
2021-04-06T16:56:24.0207313Z   asked CMake to find a package configuration file provided by "catkin", but
2021-04-06T16:56:24.0207940Z   CMake did not find one.
2021-04-06T16:56:24.0208229Z 
2021-04-06T16:56:24.0208782Z   Could not find a package configuration file provided by "catkin" with any
2021-04-06T16:56:24.0209435Z   of the following names:
2021-04-06T16:56:24.0209734Z 
2021-04-06T16:56:24.0210132Z     catkinConfig.cmake
2021-04-06T16:56:24.0210815Z     catkin-config.cmake
2021-04-06T16:56:24.0211158Z 
2021-04-06T16:56:24.0211699Z   Add the installation prefix of "catkin" to CMAKE_PREFIX_PATH or set
2021-04-06T16:56:24.0212476Z   "catkin_DIR" to a directory containing one of the above files.  If "catkin"
2021-04-06T16:56:24.0213290Z   provides a separate development package or SDK, be sure it has been
2021-04-06T16:56:24.0213882Z   installed.
2021-04-06T16:56:24.0214144Z 
2021-04-06T16:56:24.0214357Z 
2021-04-06T16:56:24.0215007Z -- Configuring incomplete, errors occurred!
2021-04-06T16:56:24.0215720Z See also "/tmp/T57Zfo/CMakeFiles/CMakeOutput.log".
2021-04-06T16:56:24.0218106Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/command_actor[0m[1;36m[0m
2021-04-06T16:56:24.0219270Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/d6ceuw[0m[1;36m[0m
2021-04-06T16:56:24.0221105Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/system_plugin[0m[1;36m[0m
2021-04-06T16:56:24.0222246Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/hY8fHb[0m[1;36m[0m
2021-04-06T16:56:24.0223371Z [1;36m[Dbg] [examples_build.cc:151] [0m[1;36mSource: [0m[1;36m/github/workspace/examples/plugin/Hello_world[0m[1;36m[0m
2021-04-06T16:56:24.0224486Z [1;36m[Dbg] [examples_build.cc:157] [0m[1;36mBuild directory: [0m[1;36m/tmp/T57Zfo[0m[1;36m[0m
2021-04-06T16:56:24.0225276Z /github/workspace/test/integration/examples_build.cc:164: Failure
2021-04-06T16:56:24.0225971Z Expected equality of these values:
2021-04-06T16:56:24.0226439Z   system(cmd)
2021-04-06T16:56:24.0226809Z     Which is: 256
2021-04-06T16:56:24.0227126Z   0
2021-04-06T16:56:24.0227777Z Hello_world
```

Also, please clean up the code and remove all references to `SampleSystem`, as well as the usage of `Hello_world` instead of `HelloWorld`.

**Comment by addy1997** [comment]
Well, sorry to have bothered you. I will do it.

**Comment by addy1997** [comment]
Following changes have been made to the files
1. References to the sample_system removes
2. rospy and roscpp dependencies (from the CMakeLists.txt file) removes
3. refactored the code to reduce clutter

I am fairly certain the CI should pass now.

**Comment by addy1997** [comment]
@chapulina, I have checked the log files it says this `[0 errors, 2 warnings]`, rest of the checks seem to pass through. There doesn't seem to be any syntax-related errors in the log file.

File: examples/plugin/Hello_world/package.xml
Comment (chapulina) [inline examples/plugin/Hello_world/package.xml:1]: Ignition doesn't use `package.xml` files, these are usually used by ROS projects. Mind removing it?

File: examples/plugin/Hello_world/HelloWorld.cc
Comment (chapulina) [inline examples/plugin/Hello_world/HelloWorld.cc:2]:
```suggestion
 * Copyright (C) 2021 Open Source Robotics Foundation
```

File: examples/plugin/Hello_world/HelloWorld.hh
Comment (chapulina) [inline examples/plugin/Hello_world/HelloWorld.hh:2]:
```suggestion
 * Copyright (C) 2021 Open Source Robotics Foundation
```

File: examples/plugin/Hello_world/README.md
Comment (chapulina) [inline examples/plugin/Hello_world/README.md:28]: Thanks for the instructions. It's now just missing a way to run it, so the user can see the string printed.

File: examples/plugin/Hello_world/HelloWorld.hh
Comment (chapulina) [inline examples/plugin/Hello_world/HelloWorld.hh:32]: Mind removing these empty lines from 25 to here?

**Review by chapulina — CHANGES_REQUESTED** [review]
Cool, the test passes now, which means the plugin compiles correctly!

Could you remove the `.so` file? We don't commit those to the repository, because they may not work for all users across all platforms.

File: examples/plugin/Hello_world/HelloWorld.hh
Comment (addy1997) [reply examples/plugin/Hello_world/HelloWorld.hh:30]: Done

File: examples/plugin/Hello_world/HelloWorld.hh
Comment (addy1997) [reply examples/plugin/Hello_world/HelloWorld.hh]: Done

File: examples/plugin/Hello_world/HelloWorld.cc
Comment (addy1997) [reply examples/plugin/Hello_world/HelloWorld.cc]: Done

File: examples/plugin/Hello_world/package.xml
Comment (addy1997) [reply examples/plugin/Hello_world/package.xml]: Done

**Comment by addy1997** [comment]
> Cool, the test passes now, which means the plugin compiles correctly!
> 
> Could you remove the `.so` file? We don't commit those to the repository, because they may not work for all users across all platforms.

@chapulina, I have removed the instructions, lib.so file, and package.xml file.

**Comment by addy1997** [comment]
@chapulina, are there any more changes required in this plugin?

**Comment by chapulina** [comment]
Hi @addy1997 , this is on my review queue :+1:

File: examples/plugin/Hello_world/CMakeLists.txt
Comment (j-rivero) [inline examples/plugin/Hello_world/CMakeLists.txt:7]: Could we use `ign_find_package` command from ign-cmake here and for ignition-gazebo4 below? IT does not make any difference right now but using the ign-cmake wrapper can provide extra benefits in the future.

Comment (addy1997) [reply examples/plugin/Hello_world/CMakeLists.txt:7]: @j-rivero, `find_package` in line 7?

Comment (addy1997) [reply examples/plugin/Hello_world/CMakeLists.txt:7]: > @j-rivero, `find_package` in line 7?

I have made the changes.

**Comment by addy1997** [comment]
@chapulina, @j-rivero, I have whittled down the code and removed the extra lines as you asked for. Can we have a final decision on this PR?

**Comment by addy1997** [comment]
@j-rivero, after making the changes, the `CI` fails which was not the case previously. Please suggest what to do now? @chapulina, do you want to merge this PR, if nothing is pending?

**Comment by addy1997** [comment]
@chapulina, please could you merge this PR?

**Review by chapulina — APPROVED** [review]
Thank you for iterating, @addy1997 ! I pushed some changes in 79bbfa93b669319321d12c2b8389111a1d0b7a57 to update style and add comments and instructions to help new users understand what's happening.

Let me know if you're ok with the changes!

**Comment by addy1997** [comment]
> Thank you for iterating, @addy1997 ! I pushed some changes in [79bbfa9](https://github.com/ignitionrobotics/ign-gazebo/commit/79bbfa93b669319321d12c2b8389111a1d0b7a57) to update style and add comments and instructions to help new users understand what's happening.
> 
> Let me know if you're ok with the changes!

Yes, I am good with the changes.

**Pull request merged**
